Improvements for boleto (fees, fines and discount) and credit card payments (3DS 2.0)

romerosilva-meli opened the pull request:
this branch adds:

- Data models for processing boletos with fees, fines and discount
- Fields for processing credit card payments with 3DS 2.0

## Checklist validación PR:

- [X] Título y descripción clara del PR
- [X] Tests de mi funcionalidad 
- [X] Documentación de mi funcionalidad
- [X] Tests ejecutados y pasados
- [X] Branch Coverage >= 80%

meliguilhermefernandes committed:
CORECH-2571 Adicionando boleto multa e juros

Commit:
Implementing 3DS.

Commit:
Adjusting 3DS atributes.

Commit:
teste

Commit:
pre-commit

Commit:
Adjusting 3DS atributes and unit tests.

Commit:
Pre-commit

Commit:
Removing unused class.

Commit:
Implementing internal_metadata.

romerosilva-meli committed:
Delete control.xml

romerosilva-meli committed:
Fix 3ds info field

to match name in response payload

romerosilva-meli committed:
Fix tests after merging commits

from boleto and 3ds branches

romerosilva-meli committed:
Bump SDK version

gdeandradero commented on src/main/java/com/mercadopago/resources/payment/Payment.java:
```diff
@@ -185,4 +185,16 @@ public class Payment extends MPResource {
+  private PaymentMethod paymentMethod;
+
+  /** 3DS Info. */
+  private PaymentThreeDSInfo threeDsinfo;
```
```suggestion
  private PaymentThreeDSInfo threeDsInfo;
```

eltinMeli replied:
É isso msm Gabs pelo que está no json de mock.

"three_dsinfo"

eltinMeli commented on src/main/java/com/mercadopago/resources/payment/Payment.java:
```diff
@@ -185,4 +185,16 @@ public class Payment extends MPResource {
+  private PaymentMethod paymentMethod;
+
+  /** 3DS Info. */
+  private PaymentThreeDSInfo threeDsinfo;
```
Tá certo esse camelCase para o retorno dessa propriedade?

eltinMeli replied:
Conferi no mock e está correta

rhames07 replied:
Para manter a congruência com a forma que estamos chamando 3DS, sempre com `DS` em maiúsculo, assim como da classe `PaymentThreeDSInfo`, não seria melhor chamar a variável de `threeDSInfo`? Realmente não sei o que seria mais correto nesse caso.

rhames07 replied:
@eltinMeli

eltinMeli commented on src/main/java/com/mercadopago/resources/payment/PaymentDiscount.java:
```diff
@@ -0,0 +1,22 @@
+  private LocalDate limitDate;
+
+  /** Amount Applied. */
+  private BigDecimal amountApplied;
```
Pode remover essa propriedade, foi removida pelo time de Core.

eltinMeli commented on src/main/java/com/mercadopago/resources/payment/PaymentFee.java:
```diff
@@ -0,0 +1,18 @@
+  private BigDecimal value;
+
+  /** Amount Applied. */
+  private BigDecimal amountApplied;
```
Pode remover essa propriedade, foi removida pelo time de Core.

eltinMeli commented on src/test/java/com/mercadopago/resources/mocks/request/payment/payment_base.json:
```diff
@@ -63,5 +63,25 @@
+          "limit_date":"2022-10-25"
+        }],
+        "fine":{
+          "type":"fixed",
```
`fixed` não é suportado para fine
```suggestion
          "type":"percentage",
```

eltinMeli commented on src/test/java/com/mercadopago/resources/mocks/request/payment/payment_base.json:
```diff
@@ -63,5 +63,25 @@
+          "value":2
+        },
+        "interest":{
+          "type":"fixed",
```
`fixed` não é suportado para interest
```suggestion
          "type":"percentage",
```

eltinMeli commented on src/test/java/com/mercadopago/client/payment/PaymentClientTest.java:
```diff
@@ -633,12 +667,36 @@ private PaymentCreateRequest newCardPayment() {
             .build();
 
+    PaymentFeeRequest fine =
+        PaymentFeeRequest.builder().type("fixed").value(new BigDecimal(2)).build();
```
`fixed` não é suportado para fine
```suggestion
        PaymentFeeRequest.builder().type("percentage").value(new BigDecimal(2)).build();
```

eltinMeli commented on src/test/java/com/mercadopago/client/payment/PaymentClientTest.java:
```diff
@@ -633,12 +667,36 @@ private PaymentCreateRequest newCardPayment() {
+        PaymentFeeRequest.builder().type("fixed").value(new BigDecimal(2)).build();
+
+    PaymentFeeRequest interest =
+        PaymentFeeRequest.builder().type("fixed").value(new BigDecimal("0.03")).build();
```
`fixed` não é suportado para interest
```suggestion
        PaymentFeeRequest.builder().type("percentage").value(new BigDecimal("0.03")).build();
```

eltinMeli commented on src/test/java/com/mercadopago/resources/mocks/request/payment/payment_3ds.json:
```diff
@@ -0,0 +1,87 @@
+        }
+      }
+    }
+  },
```
Esses dados são relacionados a boleto com multa, creio que veio junto por engano para 3ds

romerosilva-meli committed:
Remove payload related to boleto fine

from 3ds payment mock

romerosilva-meli committed:
Remove unused fields

from PaymentDiscount, PaymentFee

romerosilva-meli committed:
Change unsupported type values

'fixed' is not a valid type value for fine and interest

romerosilva-meli committed:
Add test for boleto interest, fees and discount

and refactor helper methods to create payments

romerosilva-meli committed:
Re-order fields

rhames07 commented on src/main/java/com/mercadopago/resources/payment/PaymentThreeDSInfo.java:
```diff
@@ -0,0 +1,15 @@
+
+/** 3DS Info. */
+@Getter
+@Builder
```
O objeto instanciado dessa classe deverá ser apenas de leitura para o usuário, assim como os outros arquivos dentro de `/resources`, então não deveria ter o comportamento builder, certo?
```suggestion
```

romerosilva-meli committed:
Add changes requested by reviewers

- Fix threeDSInfo field name
- Remove Builder annotation in PaymentThreeDSInfo
- Change PaymentFeeRequest type to percentage
